hydra_plugins/hydra_wandb_sweeper/config.py

```diff
@@ -1,30 +1,95 @@
-import dataclasses
-import typing
+from dataclasses import dataclass, field
+from typing import Any, Dict, MutableSequence, Optional
 
-import omegaconf
-from hydra.core import config_store
+from hydra.core.config_store import ConfigStore
+from omegaconf import DictConfig, ListConfig
 
 
-@dataclasses.dataclass
+@dataclass
+class WandbParameterSpec:
+    """Representation of all the options to define
+    a Wandb parameter.
+    """
+
+    # (int or float) Maximum and minimum values.
+    # if int, for int uniform-distributed hyperparameters. If float, for uniform-distributed hyperparameters.
+    min: Optional[float] = None
+    max: Optional[float] = None
+
+    # (float) Mean parameter for normal or log normal-distributed hyperparameters.
+    mu: Optional[float] = 0
+
+    # (float) Standard deviation parameter for normal or log normal-distributed hyperparameters.
+    sigma: Optional[float] = 1
+
+    # (float) Quantization step size for quantized hyperparameters.
+    q: Optional[float] = 1
+
+    # Specify how values will be distributed
+    # Supported distributions and required parameters
+    # "constant": ["value"],
+    # "categorical": ["values"],
+    # "int_uniform": ["min", "max"],
+    # "uniform": ["min", "max"],
+    # "q_uniform": ["min", "max", "q"],
+    # "log_uniform": ["min", "max"],
+    # "q_log_uniform": ["min", "max", "q"],
+    # "inv_log_uniform": ["min", "max"],
+    # "normal": ["mu", "sigma"],
+    # "q_normal": ["mu", "sigma", "q"],
+    # "log_normal": ["mu", "sigma"],
+    # "q_log_normal": ["mu", "sigma", "q"],
+    distribution: Optional[str] = None
+    value: Optional[float] = None
+    values: Optional[MutableSequence] = None
+
+
+@dataclass
 class WandbConfig:
-    name: str
-    method: str
-    count: typing.Optional[int] = None
-    metric: typing.Optional[omegaconf.DictConfig] = omegaconf.DictConfig({})
-    num_agents: typing.Optional[int] = 1
-    sweep_id: typing.Optional[str] = None
-    entity: typing.Optional[str] = None
-    project: typing.Optional[str] = None
-    early_terminate: typing.Optional[omegaconf.DictConfig] = omegaconf.DictConfig({})
-    preemptible: typing.Optional[bool] = False
-
-
-@dataclasses.dataclass
+    name: str = "hydra_wandb_sweep"
+    method: str = "bayes"
```

Comment:
Can we make this a required parameter instead of an optional one? Don't think setting it to bayesian makes sense.

Reply:
Oof, so this one is tricky to explain. There is a very odd Hydra-specific situation where it's necessary for this to have default values. Specifically, because of submitit pre-emption via the Hydra submitit plugin.

I agree it should be a required parameter and since I want this to be as launcher agnostic as possible, I shouldn't be making a decision such as above because of peculiarities from the launcher. I'll find a workaround.

Reply:
I forgot that OmegaConf has the `MISSING` signifier for required parameters that I could test out. It could perhaps workaround the issue I briefly mentioned. I'm taking a look at the [Optuna sweeper config as an example](https://github.com/facebookresearch/hydra/blob/796fdf1c54eabdcd5121d3a9c8ab6af70ee41763/plugins/hydra_optuna_sweeper/hydra_plugins/hydra_optuna_sweeper/config.py#L140).